src/HexagonOffload.cpp

```diff
@@ -12,13 +12,558 @@
+    R_HEX_12_X = 24,
+    R_HEX_11_X = 25,
+    R_HEX_10_X = 26,
+    R_HEX_9_X = 27,
+    R_HEX_8_X = 28,
+    R_HEX_7_X = 29,
+    R_HEX_6_X = 30,
+    R_HEX_32_PCREL = 31,
+    R_HEX_COPY = 32,
+    R_HEX_GLOB_DAT = 33,
+    R_HEX_JMP_SLOT = 34,
+    R_HEX_RELATIVE = 35,
+    R_HEX_PLT_B22_PCREL = 36,
+    R_HEX_GOTREL_LO16 = 37,
+    R_HEX_GOTREL_HI16 = 38,
+    R_HEX_GOTREL_32 = 39,
+    R_HEX_GOT_LO16 = 40,
+    R_HEX_GOT_HI16 = 41,
+    R_HEX_GOT_32 = 42,
+    R_HEX_GOT_16 = 43,
+    R_HEX_DTPMOD_32 = 44,
+    R_HEX_DTPREL_HI16 = 46,
+    R_HEX_DTPREL_32 = 47,
+    R_HEX_DTPREL_16 = 48,
+    R_HEX_GD_PLT_B22_PCREL = 49,
+    R_HEX_GD_GOT_LO16 = 50,
+    R_HEX_GD_GOT_HI16 = 51,
+    R_HEX_GD_GOT_32 = 52,
+    R_HEX_GD_GOT_16 = 53,
+    R_HEX_IE_LO16 = 54,
+    R_HEX_IE_HI16 = 55,
+    R_HEX_IE_32 = 56,
+    R_HEX_IE_GOT_LO16 = 57,
+    R_HEX_IE_GOT_HI16 = 58,
+    R_HEX_IE_GOT_32 = 59,
+    R_HEX_IE_GOT_16 = 60,
+    R_HEX_TPREL_LO16 = 61,
+    R_HEX_TPREL_HI16 = 62,
+    R_HEX_TPREL_32 = 63,
+    R_HEX_TPREL_16 = 64,
+    R_HEX_6_PCREL_X = 65,
+    R_HEX_GOTREL_32_6_X = 66,
+    R_HEX_GOTREL_16_X = 67,
+    R_HEX_GOTREL_11_X = 68,
+    R_HEX_GOT_32_6_X = 69,
+    R_HEX_GOT_16_X = 70,
+    R_HEX_GOT_11_X = 71,
+    R_HEX_DTPREL_32_6_X = 72,
+    R_HEX_DTPREL_16_X = 73,
+    R_HEX_DTPREL_11_X = 74,
+    R_HEX_GD_GOT_32_6_X = 75,
+    R_HEX_GD_GOT_16_X = 76,
+    R_HEX_GD_GOT_11_X = 77,
+    R_HEX_IE_32_6_X = 78,
+    R_HEX_IE_16_X = 79,
+    R_HEX_IE_GOT_32_6_X = 80,
+    R_HEX_IE_GOT_16_X = 81,
+    R_HEX_IE_GOT_11_X = 82,
+    R_HEX_TPREL_32_6_X = 83,
+    R_HEX_TPREL_16_X = 84,
+    R_HEX_TPREL_11_X = 85,
+    R_HEX_LD_PLT_B22_PCREL = 86,
+    R_HEX_LD_GOT_LO16 = 87,
+    R_HEX_LD_GOT_HI16 = 88,
+    R_HEX_LD_GOT_32 = 89,
+    R_HEX_LD_GOT_16 = 90,
+    R_HEX_LD_GOT_32_6_X = 91,
+    R_HEX_LD_GOT_16_X = 92,
+    R_HEX_LD_GOT_11_X = 93,
+};
+
+// This logic comes from support from Qualcomm.
+bool maybe_branch_inst(uint32_t reloc_type) {
+    switch (reloc_type) {
+    case R_HEX_PLT_B22_PCREL:
+    case R_HEX_B22_PCREL:
+    case R_HEX_B22_PCREL_X:
+    case R_HEX_B15_PCREL:
+    case R_HEX_B15_PCREL_X:
+    case R_HEX_B13_PCREL:
+    case R_HEX_B13_PCREL_X:
+    case R_HEX_B9_PCREL:
+    case R_HEX_B9_PCREL_X:
+    case R_HEX_B7_PCREL:
+    case R_HEX_B7_PCREL_X:
+    case R_HEX_B32_PCREL_X:
+    case R_HEX_32_PCREL:
+    case R_HEX_6_PCREL_X:
+
+    case R_HEX_LO16:
+    case R_HEX_HI16:
+    case R_HEX_16:
+    case R_HEX_8:
+    case R_HEX_32_6_X:
+    case R_HEX_16_X:
+    case R_HEX_12_X:
+    case R_HEX_11_X:
+    case R_HEX_10_X:
+    case R_HEX_9_X:
+    case R_HEX_8_X:
+    case R_HEX_7_X:
+    case R_HEX_6_X:
+    case R_HEX_32:
+        return true;
+    default:
+        return false;
+    }
+}
+
+std::string hex(uint32_t x) {
+    char buffer[100];
+    snprintf(buffer, sizeof(buffer), "0x%08x", x);
+    return buffer;
+}
+
+void do_reloc(char *addr, uint32_t mask, uintptr_t val, bool is_signed, bool verify) {
```

Comment:
Do you mean that this relocation handling code is now available in LLVM? A quick googling hasn't turned up the relevant code yet, but we would definitely prefer to depend on LLVM to do this... Ideally, we would use LLVM to do the whole linker job, we would love to get rid of this custom linker if possible.